Bukkit/0087-Add-PlayerSettingsUpdateEvent.patch

```diff
@@ -0,0 +1,389 @@
+@@ -15,6 +15,7 @@ import org.bukkit.Material;
+ import org.bukkit.Note;
+ import org.bukkit.OfflinePlayer;
+ import org.bukkit.Particle;
++import org.bukkit.PlayerSettings;
+ import org.bukkit.Skin;
+ import org.bukkit.Sound;
+ import org.bukkit.Statistic;
+@@ -1643,4 +1644,28 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
+      * If the given item type is not cooling down, 0 is returned.
+      */
+     float getRemainingItemCooldown(Material item);
++
++    /**
++     * Get the various client-established settings of this player.
++     *
++     * If the player has not sent these values to the server,
++     * each getter, with the exclusion of {@link #getLocale()},
++     * will return null.
++     *
++     * @return the player's client settings.
++     */
++    PlayerSettings getSettings();
++
++    /**
++     * Forcibly set the player's client-established settings.
++     *
++     * This is recommended if the server has previous knowledge
++     * of a player's settings (like being stored in a database).
++     *
++     * Unlike when the player sends the packet to the server,
++     * this method does not call {@link org.bukkit.event.player.PlayerSettingsUpdateEvent}.
++     *
++     * @param settings the new player settings.
++     */
++    void setSettings(PlayerSettings settings);
+ }
+diff --git a/src/main/java/org/bukkit/event/player/PlayerSettingsUpdateEvent.java b/src/main/java/org/bukkit/event/player/PlayerSettingsUpdateEvent.java
+new file mode 100644
+index 0000000..601b0ce
+--- /dev/null
++++ b/src/main/java/org/bukkit/event/player/PlayerSettingsUpdateEvent.java
+@@ -0,0 +1,122 @@
++package org.bukkit.event.player;
++
++import com.google.common.collect.ImmutableSet;
++import org.apache.commons.lang.builder.EqualsBuilder;
++import org.bukkit.ChatVisibility;
++import org.bukkit.PlayerSettings;
++import org.bukkit.Skin;
++import org.bukkit.entity.Player;
++import org.bukkit.event.HandlerList;
++import org.bukkit.inventory.MainHand;
++
++/**
++ * Called when a {@link Player} makes changes to their {@link PlayerSettings}.
++ *
++ * {@link PlayerLocaleChangeEvent}, {@link PlayerChangedMainHandEvent},
++ * and {@link PlayerSkinPartsChangeEvent} are also fired if their respective
++ * fields were changed.
++ *
++ */
++public class PlayerSettingsUpdateEvent extends PlayerEvent {
++
++    private final static HandlerList handlers = new HandlerList();
++
++    private final PlayerSettings old;
++    private final PlayerSettings current;
++
++    public PlayerSettingsUpdateEvent(Player who, PlayerSettings old, PlayerSettings current) {
++        super(who);
++        this.old = old;
++        this.current = current;
++    }
++
++    public boolean hasUpdated() {
++        return old != null;
++    }
++
++    public PlayerSettings getOldUpdate() {
++        return old;
++    }
++
++    public PlayerSettings getNewUpdate() {
++        return current;
++    }
++
++    public String getNewLocale() {
```

Comment:
We don't need all these delegate methods, just a way to get the old and new `PlayerSettings`. The caller can easily check all this other stuff.